ugc/serdes_json.hpp

```diff
@@ -0,0 +1,226 @@
+#pragma once
+
+#include "ugc/serdes.hpp"
+#include "ugc/types.hpp"
+
+#include "coding/multilang_utf8_string.hpp"
+#include "coding/point_to_integer.hpp"
+#include "coding/reader.hpp"
+#include "coding/varint.hpp"
+#include "coding/write_to_sink.hpp"
+
+#include "base/exception.hpp"
+
+#include "3party/jansson/myjansson.hpp"
+
+#include <cmath>
+#include <cstdint>
+#include <iostream>
+
+namespace ugc
+{
+  template <typename Sink>
+  class SerializerJson
+  {
+  public:
+    SerializerJson(Sink & sink, HeaderV0 const & header) : m_sink(sink), m_header(header) {}
+
+    void operator()(bool const d, char const * name = nullptr)
+    {
+      ToJSONObject(*m_json, name, d);
+    }
+
+    void operator()(uint8_t const d, char const * name = nullptr)
+    {
+      ToJSONObject(*m_json, name, d);
+    }
+
+    void operator()(uint32_t const d, char const * name = nullptr)
+    {
+      ToJSONObject(*m_json, name, d);
+    }
+
+    void operator()(uint64_t const d, char const * name = nullptr)
+    {
+      ToJSONObject(*m_json, name, d);
+    }
+
+    void operator()(std::string const & s, char const * name = nullptr)
+    {
+      ToJSONObject(*m_json, name, s);
+    }
+
+    void operator()(Time const & t, char const * name = nullptr)
+    {
+      (*this)(ToDaysSinceEpoch(t), name);
+    }
+
+    void operator()(Sentiment sentiment, char const * name = nullptr)
+    {
+      switch (sentiment)
+      {
+        case Sentiment::Negative: return (*this)(false, name);
+        case Sentiment::Positive: return (*this)(true, name);
+      }
+    }
+
+    template <typename T>
+    void operator()(vector<T> const & vs, char const * name = nullptr)
+    {
+      my::JSONPtr safe_json = std::move(m_json);
+      m_json = my::NewJSONArray();
+      for (auto const & v : vs)
+        (*this)(v);
+      Update(std::move(safe_json), name);
+    }
+
+    template <typename R>
+    void operator()(R const & r, char const * name = nullptr)
+    {
+      my::JSONPtr safe_json = std::move(m_json);
+      m_json = my::NewJSONObject();
```

Comment:
Same as below - is it possible to put all these safe_json saving and restoring into a helper function?